src/Microsoft.OData.Client/Materialization/MaterializerAnnotationsCache.cs

```diff
@@ -0,0 +1,66 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="MaterializerAnnotationsCache.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Microsoft.OData.Client.Materialization
+{
+    /// <summary>
+    /// A cache used to store temporary materialization metadata
+    /// for deseriliazed <see cref="ODataItem"/>s during materialization of response payloads
+    /// into client CLR objects. Keys are identified using reference equality.
+    /// </summary>
+    /// <remarks>
+    /// Instances of the cache are not thread-safe. Each instance is expected to be used within the
+    /// scope of a single request.
+    /// 
+    /// The cache is designed to store one entry type per key. For example, if the key is
+    /// an <see cref="ODataResource"/>, its value should be a <see cref="MaterializerEntry"/>. It also
+    /// expects the value for any given key will not change.
+    /// So you should call `cache.SetAnnotation(odataResource, materializerEntry)` add the cache entry
+    /// and `cache.GetAnnotations<MaterializerEntry>(odataResource)` to retrieve the entry.
+    /// 
+    /// If the requirements change such that we need to allow updating the value of a key to a different type,
+    /// then the cache should be redesigned.
+    /// </remarks>
+    internal sealed class MaterializerAnnotationsCache

```

Comment:
I'd prefer `MaterializerCache` or `ODataMaterializerCache`

Reply:
I don't have an objection to the name, but I feel like this cache is used for a very specific and limited use-case, as opposed to a general purpose cache.

Reply:
I have renamed it to `MaterializerCache` and renamed the parameters accordingly.